### ISSUE TITLE
[API Proposal]: Creating linked cancellation tokens hides which token triggered cancellation

### ISSUE THREAD

**Issue by ygoe** [dotnet/runtime#78880]
### Background and motivation

See #22172. I have the same problem and just ran against it. I understood what @davidfowl was describing in the very first post. Let me add another example of what happens here:

```cs
async Task Connect(CancellationToken cancellationToken)
{
	// Stop connecting when requested, but at last after 5 seconds
	using var timeoutCts = new CancellationTokenSource(5000);
	using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutToken);
	await tcpClient.ConnectAsync("localhost", 1234, linkedCts.Token);
}

try
{
	await Connect(token);
	// Now cancel that token
}
catch (OperationCanceledException ex) when (ex.CancellationToken == token)
{
	// Should end up here, after all, we've cancelled this very token and that was the cause to return
	Console.WriteLine("Cancelled by user");
}
```

With this, when `token` is cancelled, I'd expect to see the message "Cancelled by user". But I don't. Because `token` isn't what triggered the cancellation exception. It's the linked token with the added timeout.

As the user of that API, I have no clue that it internally links the provided token with something else. When I give it a token and cancel it, I expect this to be the cancellation cause, not something else I don't know (the internal linked token).

### API Proposal

I have no idea how to solve this. All I can say is that something is missing here.

### API Usage

see above

**Comment by svick** [comment]
I think the question this proposal needs to answer is (https://github.com/dotnet/runtime/issues/22172 never managed it): Why is something like `ex.CancellationToken.IsDerivativeOf(token)` better than just `token.IsCancellationRequested`?

**Comment by ygoe** [comment]
Well, okay, I can do that. But my impression is that the `OperationCanceledException.CancellationToken` property is closer to the source. By the time the exception is handled, things might have changed and other tokens might have triggered, too (especially with timeouts). If directly checking the tokens I have is just as good, then what use does the exception's property have? It's a little bit confusing.

**Comment by stephentoub** [comment]
It's the responsibility of the API taking the cancellation token to throw an exception containing that cancellation token if it may internally have been monitoring something else based on it (and if it wants to hide that detail).  You have this as an example:
```C#
async Task Connect(CancellationToken cancellationToken)
{
    // Stop connecting when requested, but at last after 5 seconds
    using var timeoutCts = new CancellationTokenSource(5000);
    using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutToken);
    await tcpClient.ConnectAsync("localhost", 1234, linkedCts.Token);
}
```
but if you want callers of your method to be able to use the resulting OCE.Token and compare it to the token passed in, it should instead be written as something like this:
```C#
async Task Connect(CancellationToken cancellationToken)
{
    // Stop connecting when requested, but at last after 5 seconds
    CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
    linkedCts.CancelAfter(5000);
    try
    {
        await tcpClient.ConnectAsync("localhost", 1234, linkedCts.Token);
    }
    catch (OperationCanceledException oce) when (cancellationToken.IsCancellationRequested)
    {
        cancellationToken.ThrowIfCancellationRequested();
    }
    finally
    {
        linkedCts.Dispose();
    }
}
```
In other words, the API is putting up a contract around accepting a cancellation token, and if that token has cancellation requested, the resulting exception will appropriately contain that token.  As an implementation detail, the API is actually creating some other token and passing that along, so it's also responsible for reversing that transformation on the way out (if it wants to hide that implementation detail).

But, to @svick's point, if all you care about is whether your token is canceled, then that's all you need to check.  It generally doesn't really matter what the exact thing was the implementation noticed that caused it to cancel; if there are multiple signals that race to trigger cancellation, then any of them can reasonably be blamed as the cause.

**Comment by ygoe** [comment]
Alright, based on your suggestion, I think this should work:

```cs
using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
linkedCts.CancelAfter(5000);
try
{
	await tcpClient.ConnectAsync("localhost", 1234, linkedCts.Token);
	// do something with the connection, still subject to the timeout
}
catch (OperationCanceledException ex) when (!cancellationToken.IsCancellationRequested)
{
	// Throw a specialised exception
	throw new TimeoutException("The operation has timed out.", ex);
}
catch (OperationCanceledException ex)
{
	// Rewrite the exception to include the provided CancellationToken, preserving
	// the full stack trace to the cancelled operation in the inner exception
	throw new OperationCanceledException("The operation was canceled.", ex, cancellationToken);
}
```

I didn't know the `CancelAfter` method. It seems to avoid one extra `CancellationTokenSource` instance I had used.

It replaces a timeout-caused exception by a `TimeoutException` and rewrites another `OperationCanceledException` to contain the token that was provided by the caller. Anything else is thrown untouched. It also preserves the original stack trace in any case which may be helpful for debugging.

The only downsides are that I had to duplicate the original exception messages, but they're not critical or long anyway. And other OCEs thrown by the called methods (maybe from a global shutdown CTS) are misinterpreted as timeout.

Maybe I can wrap this in some helper code for easy reuse. This is the best I could do:

```cs
private static CancellationTokenSource MakeTimeout(CancellationToken cancellationToken, TimeSpan timeout)
{
	var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
	linkedCts.CancelAfter(timeout);
	return linkedCts;
}

private static Exception GetOperationCanceledOrTimeoutException(OperationCanceledException ex, CancellationToken cancellationToken)
{
	if (!cancellationToken.IsCancellationRequested)
	{
		// Throw a specialised exception
		return new TimeoutException("The operation has timed out.", ex);
	}
	else
	{
		// Rewrite the exception to include the provided CancellationToken, preserving
		// the full stack trace to the cancelled operation in the inner exception
		return new OperationCanceledException("The operation was canceled.", ex, cancellationToken);
	}
}
```

```cs
using var linkedCts = MakeTimeout(cancellationToken, 5000);
object something;
try
{
	await tcpClient.ConnectAsync("localhost", 1234, linkedCts.Token);
	// do something with the connection, still subject to the timeout
	something = something();
}
catch (OperationCanceledException ex)
{
	throw GetOperationCanceledOrTimeoutException(ex, cancellationToken);
}
use(something);
```

The final `throw` statement is necessary so that I can continue with previously unset variables (like `something`) afterwards. A method (similar to `ThrowIfCancellationRequested`) that will always throw is not enough.

**Comment by stephentoub** [comment]
> The only downsides are that I had to duplicate the original exception messages, but they're not critical or long anyway.

You can't use ex.Message?

**Comment by stephentoub** [comment]
> Mh, yes, for the same type. TimeoutException usually uses another message.

Sure, but you're choosing to construct a new exception type unrelated to OperationCanceledException.  Creating a completely different exception isn't something your original issue would address, either.

**Comment by ygoe** [comment]
>  Creating a completely different exception isn't something your original issue would address, either.

True. It has evolved. ;-) Since handling timeouts is different from handling requested cancellations (at least in my case, with retries and for logging), this same exception type has bothered my anyway.

So all that's left for this issue is probably tracking which CancellationToken has initially caused the cancellation. For a linked token, I guess there could be an additional property besides `IsCancellationRequested` that indicates which token that originated from. And the OCE would then go down that route and use the deepest original token it finds (there can only be one).

If I were to design this API, it could look like this:

```cs
struct CancellationToken
{
    public CancellationToken OriginalCancellationToken { get; }
}
```

**Comment by stephentoub** [comment]
> So all that's left for this issue is probably tracking which CancellationToken has initially caused the cancellation.

My point has been that I don't believe this is desirable or needed. :smile:

But regardless, https://github.com/dotnet/runtime/issues/69556 tracks that, so I'm going to close this as a duplicate.

Thanks.
